pyrobosim/pyrobosim/navigation/astar_grid.py

```diff
@@ -0,0 +1,164 @@
+
+
+class AstarGrid:
+    """Occupancy grid based A-star planner"""
+
+    def __init__(self, w, resolution=0.05, inflation_radius=0.0) -> None:
+        self.grid = occupancy_grid_from_world(
+            w, resolution=resolution, inflation_radius=inflation_radius
+        )
+        self.world = w
+        self.resolution = resolution
+        self.inflation_radius = inflation_radius
+        self.candidates = []  # Node
+        self.visited = []  # (x, y)
+        self.start = None
+        self.goal = None
+
+    def _reset(self):
+        """Resets the state of data used by the algorithm"""
+        self.grid = occupancy_grid_from_world(
+            self.world,
+            resolution=self.resolution,
+            inflation_radius=self.inflation_radius,
+        )
+        self.candidates.clear()
+        self.visited.clear()
+        self.goal = None
+        self.start = None
+        self.latest_path = []
+
+    def _heuristic(self, p1, p2):
+        """
+        Computes a heuristc between 2 points
+        :param p1: Source point
+        :type p1: Tuple (x, y)
+        :param p2: destination point
+        :type p2: Tuple (x, y)
+        """
+        return abs(p1[0] - p2[0]) + abs(p1[1] - p2[1])
+
+    def _expand(self, node):
+        """Generates and adds free neighbours to exploration candidates"""
+        motions = [(-1, 0), (1, 0), (0, 1), (0, -1), (-1, 1), (-1, -1), (1, 1), (1, -1)]
+        motion_costs = [1, 1, 1, 1, 1.414, 1.414, 1.414, 1.414]
+        for i in range(len(motions)):
+            _x = node.x + motions[i][0]
+            _y = node.y + motions[i][1]
+            if (_x, _y) not in self.visited and not self.grid.is_occupied(_x, _y):
+                _g = node.g + motion_costs[i]
+                _h = self._heuristic((_x, _y), self.goal)
+                self.candidates.append(Node(_x, _y, _g, _h, parent=node))
+
+    def _rank_candidates(self):
+        """Orders the available candidates by their cost `f`"""
+        # TODO : Can this be optimized to avoid sorting the actual list ?
+        self.candidates = sorted(self.candidates, key=lambda node: node.f, reverse=True)
+
+    def _get_best_candidate(self):
+        """Return the candidate with best metric"""
+        self._rank_candidates()
+        return self.candidates.pop()
+
+    def _mark_visited(self, node):
+        """
+        Adds the given node to visited nodes collection
+        :param node: The node that has been visited
+        :type node: :class: Node
+        """
+        self.visited.append((node.x, node.y))
+
+    def _generate_path(self, node):
+        """
+        Backtrack from a node to generate the full path till that node
+        """
+        poses = []
+        while node is not None:
+            x, y = self.grid.grid_to_world(node.x, node.y)
+            poses.append(Pose(x=x, y=y))
+            node = node.parent
+        poses.reverse()
+        self.latest_path = Path(poses=poses)
+        self.latest_path.fill_yaws()
```

Comment:
I wonder if instead of using `fill_yaws` after the fact, you could calculate this as you expand the graph because you know the direction the robot is coming from? Shouldnt be a big deal though.

In fact, this is probably more efficient this way because you only calculate yaw for the nodes along the resulting path.

Reply:
Calculating it at the end is the best in my opinion as well.
Since yaws for nodes that are not on the path are not useful elsewhere.